lib/logitech_receiver/device.py

```diff
@@ -0,0 +1,357 @@
+                # if unifying else 0x03)
+                kind = ord(link_notification.data[0:1]) & 0x0F
+                # fix EX100 wpid
+                if receiver.ex100_wpid_fix:  # EX100 receiver
+                    self.wpid = _strhex(link_notification.data[2:3]) + '00'
+                    # workaround for EX100 switched kind
+                    if self.wpid == '3F00':
+                        kind = 2
+                    if self.wpid == '6500':
+                        kind = 1
+                self._kind = _hidpp10.DEVICE_KIND[kind]
+            else:
+                # force a reading of the wpid
+                pair_info = self.receiver.read_register(_R.receiver_info, 0x20 + number - 1)
+                if pair_info:
+                    # may be either a Unifying receiver, or an Unifying-ready
+                    # receiver
+                    self.wpid = _strhex(pair_info[3:5])
+                    kind = ord(pair_info[7:8]) & 0x0F
+                    self._kind = _hidpp10.DEVICE_KIND[kind]
+                elif receiver.ex100_wpid_fix:
+                    # ex100 receiver, fill fake device_info with known wpid's
+                    # accordingly to drivers/hid/hid-logitech-dj.c
+                    # index 1 or 2 always mouse, index 3 always the keyboard,
+                    # index 4 is used for an optional separate numpad
+                    if number == 1:  # mouse
+                        self.wpid = '3F00'
+                        self._kind = _hidpp10.DEVICE_KIND[2]
+                    elif number == 3:  # keyboard
+                        self.wpid = '6500'
+                        self._kind = _hidpp10.DEVICE_KIND[1]
+                    else:  # unknown device number on EX100
+                        _log.error('failed to set fake EX100 wpid for device %d of %s', number, receiver)
+                        raise _base.NoSuchDevice(number=number, receiver=receiver, error='Unknown EX100 device')
+                else:
+                    # unifying protocol not supported, must be a Nano receiver
+                    device_info = self.receiver.read_register(_R.receiver_info, 0x04)
+                    if device_info is None:
+                        _log.error('failed to read Nano wpid for device %d of %s', number, receiver)
+                        raise _base.NoSuchDevice(number=number, receiver=receiver, error='read Nano wpid')
+
+                    self.wpid = _strhex(device_info[3:5])
+                    self._power_switch = '(' + _('unknown') + ')'
+
+            # the wpid is necessary to properly identify wireless link on/off
+            # notifications also it gets set to None on this object when the
+            # device is unpaired
+            assert self.wpid is not None, 'failed to read wpid: device %d of %s' % (number, receiver)
+
+            for dev in _hid.enumerate({'vendor_id': 0x046d, 'product_id': int(self.receiver.product_id, 16)}):
```

Comment:
Can you print the `dev.serial` in this loop (and post the output), it seems like unifying devices structure the string differently, we might need to use the `HID_ID` instead if it turns out to be inconsistent. @pfps

Reply:
I added `print("ENUMERATE",dev.serial,dev)`  Here is the result

```
ENUMERATE 1025-93-e8-52-7f DeviceInfo(path='/dev/hidraw1', vendor_id='046d', product_id='c52b', serial='1025-93-e8-52-7f', release=b'1210', manufacturer=b'Logitech', product=b'USB Receiver', interface=2, driver='logitech-hidpp-device')
ENUMERATE 1025-93-e8-52-7f DeviceInfo(path='/dev/hidraw1', vendor_id='046d', product_id='c52b', serial='1025-93-e8-52-7f', release=b'1210', manufacturer=b'Logitech', product=b'USB Receiver', interface=2, driver='logitech-hidpp-device')
ENUMERATE 4051-60-e4-b9-7d DeviceInfo(path='/dev/hidraw5', vendor_id='046d', product_id='c52b', serial='4051-60-e4-b9-7d', release=b'1210', manufacturer=b'Logitech', product=b'USB Receiver', interface=2, driver='logitech-hidpp-device')
```

Reply:
This might be related to another problem though.   There are two protocols involved with Logitech devices - HID++ and DJ.   The notifications for them are different - to the point that DJ pairing notifications were producing "fake" devices with different WPIDs.  Of course this was all timing-dependent because the processing of the notifications could happen in different orders.  What I did (and this was only very recently added to Solaar) was to ignore DJ pairing notifications.

Reply:
hm that is strange, are you able to use the "show" command?

Reply:
show works fine, with the same ENUMERATE output.  Note that I have two mice on the receiver, thus the three outputs (one for when looking for the first device, two when looking for the second).

Reply:
Ah, I haven't touched anything involving notifications yet.

Reply:
It may be that you don't have to worry about the DJ notifications at all.  The recent changes mean that these fake devices don't show up in Solaar any more.